File: src/algorithms/linked-lists/insertion-deletion/remove-duplicates-sorted/remove-duplicates-sorted.test.ts
import { describe, it, expect } from "vitest";
import { removeDuplicatesSorted } from "./sources/remove-duplicates-sorted.ts?fn";

interface ListNode {
  value: number;
  next: ListNode | null;
}

function buildList(values: number[]): ListNode | null {
  let head: ListNode | null = null;
  for (let idx = values.length - 1; idx >= 0; idx--) {
    head = { value: values[idx]!, next: head };
  }
  return head;
}

function listToArray(head: ListNode | null): number[] {
  const result: number[] = [];
  let current: ListNode | null = head;
  while (current !== null) {
    result.push(current.value);
    current = current.next;
  }
  return result;
}

describe("removeDuplicatesSorted", () => {
  it("removes consecutive duplicates from a sorted list", () => {
    const result = removeDuplicatesSorted(buildList([1, 1, 2, 3, 3, 3, 4, 5, 5]));
    expect(listToArray(result as ListNode | null)).toEqual([1, 2, 3, 4, 5]);
  });

  it("leaves a list with no duplicates unchanged", () => {
    const result = removeDuplicatesSorted(buildList([1, 2, 3, 4, 5]));
    expect(listToArray(result as ListNode | null)).toEqual([1, 2, 3, 4, 5]);
  });

  it("handles a list of all duplicate values", () => {
    const result = removeDuplicatesSorted(buildList([7, 7, 7, 7]));
    expect(listToArray(result as ListNode | null)).toEqual([7]);
  });

  it("returns null for an empty list", () => {
    const result = removeDuplicatesSorted(null);
    expect(result).toBeNull();
  });

  it("handles a single-element list", () => {
    const result = removeDuplicatesSorted(buildList([5]));
    expect(listToArray(result as ListNode | null)).toEqual([5]);
  });

  it("removes duplicates from a two-element list", () => {
    const result = removeDuplicatesSorted(buildList([3, 3]));
    expect(listToArray(result as ListNode | null)).toEqual([3]);
  });

  it("keeps two different elements unchanged", () => {
    const result = removeDuplicatesSorted(buildList([1, 2]));
    expect(listToArray(result as ListNode | null)).toEqual([1, 2]);
  });

  it("removes duplicates with mixed run lengths", () => {
    const result = removeDuplicatesSorted(buildList([1, 2, 2, 3, 3, 3, 4]));
    expect(listToArray(result as ListNode | null)).toEqual([1, 2, 3, 4]);
  });
});

File: src/algorithms/linked-lists/insertion-deletion/remove-duplicates-sorted/__tests__/remove-duplicates-sorted_test.go
package main

import (
    "reflect"
    "testing"
)

func buildListRemoveDuplicatesSorted(values []int) *ListNode {
    var head *ListNode
    for idx := len(values) - 1; idx >= 0; idx-- {
        head = &ListNode{value: values[idx], next: head}
    }
    return head
}

func listToSliceRemoveDuplicatesSorted(head *ListNode) []int {
    result := []int{}
    for head != nil {
        result = append(result, head.value)
        head = head.next
    }
    return result
}

func TestRemoveDuplicatesSortedConsecutive(t *testing.T) {
    result := removeDuplicatesSorted(buildListRemoveDuplicatesSorted([]int{1, 1, 2, 3, 3, 3, 4, 5, 5}))
    if !reflect.DeepEqual(listToSliceRemoveDuplicatesSorted(result), []int{1, 2, 3, 4, 5}) {
        t.Error("expected [1 2 3 4 5] after removing consecutive duplicates")
    }
}

func TestRemoveDuplicatesSortedNoDuplicates(t *testing.T) {
    result := removeDuplicatesSorted(buildListRemoveDuplicatesSorted([]int{1, 2, 3, 4, 5}))
    if !reflect.DeepEqual(listToSliceRemoveDuplicatesSorted(result), []int{1, 2, 3, 4, 5}) {
        t.Error("expected list unchanged when no duplicates")
    }
}

func TestRemoveDuplicatesSortedAllSame(t *testing.T) {
    result := removeDuplicatesSorted(buildListRemoveDuplicatesSorted([]int{7, 7, 7, 7}))
    if !reflect.DeepEqual(listToSliceRemoveDuplicatesSorted(result), []int{7}) {
        t.Error("expected [7] for all-duplicate list")
    }
}

func TestRemoveDuplicatesSortedEmptyList(t *testing.T) {
    result := removeDuplicatesSorted(nil)
    if result != nil {
        t.Error("expected nil for empty list")
    }
}

func TestRemoveDuplicatesSortedSingleElement(t *testing.T) {
    result := removeDuplicatesSorted(buildListRemoveDuplicatesSorted([]int{5}))
    if !reflect.DeepEqual(listToSliceRemoveDuplicatesSorted(result), []int{5}) {
        t.Error("expected [5] for single-element list")
    }
}

func TestRemoveDuplicatesSortedTwoElementDuplicates(t *testing.T) {
    result := removeDuplicatesSorted(buildListRemoveDuplicatesSorted([]int{3, 3}))
    if !reflect.DeepEqual(listToSliceRemoveDuplicatesSorted(result), []int{3}) {
        t.Error("expected [3] after deduplication of [3 3]")
    }
}

func TestRemoveDuplicatesSortedTwoDifferentElements(t *testing.T) {
    result := removeDuplicatesSorted(buildListRemoveDuplicatesSorted([]int{1, 2}))
    if !reflect.DeepEqual(listToSliceRemoveDuplicatesSorted(result), []int{1, 2}) {
        t.Error("expected [1 2] unchanged for two different elements")
    }
}

func TestRemoveDuplicatesSortedMixedRunLengths(t *testing.T) {
    result := removeDuplicatesSorted(buildListRemoveDuplicatesSorted([]int{1, 2, 2, 3, 3, 3, 4}))
    if !reflect.DeepEqual(listToSliceRemoveDuplicatesSorted(result), []int{1, 2, 3, 4}) {
        t.Error("expected [1 2 3 4] after mixed run deduplication")
    }
}
